components/odh-notebook-controller/controllers/notebook_webhook.go

```diff
@@ -227,6 +230,29 @@ func InjectOAuthProxy(notebook *nbv1.Notebook, oauth OAuthConfig) error {
 	return nil
 }
 
+func (w *NotebookWebhook) ClusterWideProxyIsEnabled() bool {
+	proxyResourceList := &configv1.ProxyList{}
+	err := w.Client.List(context.TODO(), proxyResourceList)
+	if err != nil {
+		return false
+	}
+
+	for _, proxy := range proxyResourceList.Items {
+		if proxy.Name == "cluster" {
+			if proxy.Status.HTTPProxy != "" && proxy.Status.HTTPSProxy != "" &&
+				proxy.Status.NoProxy != "" {
+				// Update Proxy Env variables map
+				proxyEnvVars["HTTP_PROXY"] = proxy.Status.HTTPProxy
```

Comment:
> We need to postpone the env vars update to when the notebook is restarted by the user, to prevent unexpected restart.

If that is really necessary, I would not know how to achieve this. In my view, it is overkill, but I realize you sometimes have differing requirements with environments with long running notebook containers.

Reply:
I'm setting up a fake operator upgrade scenario with envtest and so far it looks like that changing the envvars here does not modify the existing statefulsets. That seems to be because webhooks run only on create, update (and delete) of the notebook spec, and there's nothing else changing the spec, so the webhook is not rerun when operator restarts.

We have that one Openshift 4.13 cluster that does recreate pods where other clusters would not, so we need to figure out that case, but I'm hopeful this can be done without restarts somehow.

Reply:
Here are two proposed solutions to the notebook restart problem due to updates to the values that odh-notebook-controller is setting in the webhook

* https://github.com/opendatahub-io/kubeflow/pull/436
* https://github.com/opendatahub-io/kubeflow/pull/437

Reply:
@jiridanek is this still an issue? the restart issue. I lost track a little over time.

Reply:
This should be fixed by 

* https://github.com/opendatahub-io/kubeflow/pull/436

> Notebook CRs have a new annotation `notebooks.opendatahub.io/update-pending` that is added when there is a mismatch between the currently running Workbench and the desired configuration, and applying the desired configuration requires restarting the Workbench. Only one configuration change is displayed, even though there may be more than one pending updates. When there are no pending updates, for example when the Workbench is restarted and pending updates get applied on that occasion, the annotation is again removed by the odh-notebook-controller webhook.
>
> Example:

```
notebooks.opendatahub.io/update-pending: '{v1.PodSpec}.Containers[1].Image: registry.redhat.io/openshift4/ose-oauth-proxy@sha256:4f8d66597feeb32bb18699326029f9a71a5aca4a57679d636b876377c2e95695 != registry.redhat.io/openshift4/ose-oauth-proxy@sha256:4bef31eb993feb6f1096b51b4876c65a6fb1f4401fee97fa4f4542b6b7c9bc46'
```

> https://issues.redhat.com/browse/RHOAIENG-15335

There isn't a good automated test for this scenario yet, so this upgrade is one of the things I plan to try by hand, to be sure (I proposed https://github.com/opendatahub-io/kubeflow/pull/430 as automated test for upgrades, but the size of the accompanying YAMLs there scared everyone, so this is not going to be a way to do it).

Reply:
@jiridanek @jstourac I think you recently added changes to the webhook that take care of the issue.
https://github.com/opendatahub-io/kubeflow/commit/1a1894b8642f7b31492bcef07c6e626a5032fe76
So closing this as resolved.